DistinguishUnitaries/ReferenceImplementation.qs

```diff
@@ -71,7 +91,156 @@ namespace Quantum.Kata.DistinguishUnitaries {
+                Controlled unitary(qs[0..0], qs[1]);
+            }
+
+            // 0 means it was Y
+            return M(qs[0]) == Zero ? 0 | 1;
+        }
+    }
+    
+
+    operation Oracle_Reference (U : (Qubit => Unit is Adj + Ctl), power : Int, target : Qubit[]) : Unit is Adj + Ctl {
+        for (i in 1 .. power) {
+            U(target[0]);
+        }
+    }
+
+
+    // Task 1.8. Y, XZ, -Y or -XZ?
+    // Output: 0 if the given operation is the Y gate,
+    //         1 if the given operation is the -XZ gate,
+    //         2 if the given operation is the -Y gate,
+    //         3 if the given operation is the XZ gate.
+    operation DistinguishYfromXZWithPhases_Reference (unitary : (Qubit => Unit is Adj+Ctl)) : Int {
+        // Run phase estimation on the unitary and the +1 eigenstate of the Y gate |0⟩ + i|1⟩
+
+        // Construct a phase estimation oracle from the unitary
+        let oracle = DiscreteOracle(Oracle_Reference(unitary, _, _));
+
+        // Allocate qubits to hold the eigenstate of U and the phase in a big endian register 
+        mutable phaseInt = 0;
+        using ((eigenstate, phaseRegister) = (Qubit[1], Qubit[2])) {
+            let phaseRegisterBE = BigEndian(phaseRegister);
+            // Prepare the eigenstate of U
+            H(eigenstate[0]); 
+            S(eigenstate[0]);
+            // Call library
+            QuantumPhaseEstimation(oracle, eigenstate, phaseRegisterBE);
+            // Read out the phase
+            set phaseInt = MeasureInteger(BigEndianAsLittleEndian(phaseRegisterBE));
+
+            ResetAll(eigenstate);
+            ResetAll(phaseRegister);
+        }
+
+        // Convert the measured phase into return value
+        return phaseInt;
+    }
+
+
+    // ------------------------------------------------------
+    internal function ComputeRepetitions(angle : Double, offset : Int, accuracy : Double) : Int {
+        mutable pifactor = 0;
+        while (true) {
+            let pimultiple = PI() * IntAsDouble(2 * pifactor + offset);
+            let times = Round(pimultiple / angle);
+            if (AbsD(pimultiple - (IntAsDouble(times) * angle)) / PI() < accuracy) {
+                return times;
+            }
+            set pifactor += 1;
+        }
+        return 0;
+    }
+
+    // Task 1.9. Rz or Ry (fixed angle)?
+    // Output: 0 if the given operation is the Rz(θ) gate,
+    //         1 if the given operation is the Ry(θ) gate.
+    operation DistinguishRzFromRy_Reference (theta : Double, unitary : (Qubit => Unit is Adj+Ctl)) : Int {
+        using (q = Qubit()) {
+            let times = ComputeRepetitions(theta, 1, 0.1);
+            mutable attempt = 1;
+            mutable measuredOne = false;
+            repeat {
+                for (_ in 1..times) {
+                    unitary(q);
+                }
+                // for Rz, we'll never venture away from |0⟩ state, so as soon as we got |1⟩ we know it's not Rz
+                if (MResetZ(q) == One) {
```

Comment:
```suggestion
                if (M(q) == One) {
```
Would that work with the recent changes?

Reply:
Logically this one is in a loop, we're reusing the qubit and want to return it to 0 before reuse. Technically it should work, since once we get 1, we'll be breaking out from the loop, but logically I think `MResetZ` is clearer.

Reply:
That makes sense. Thanks for the explanation.